### PR TITLE
Fix non-posix paths with file:// protocol on Windows

### PR DESCRIPTION
**Pull request by fredrikaverpil**
# Pull Request Check List

Resolves: #2744 

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: poetry/installation/chef.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def get_cached_archives_for_link(self, link):  # type: (Link) -> List[Link]
         links = []
         for archive_type in archive_types:
             for archive in cache_dir.glob("*.{}".format(archive_type)):
-                links.append(Link("file://{}".format(str(archive))))
+                links.append(Link("file://{}".format(str(archive.as_posix()))))
```
Comment (abn) [inline poetry/installation/chef.py:84]:
```suggestion
                links.append(Link(archive.as_uri()))
```

**Comment by fredrikaverpil** [comment]
@abn hm, the errors I am getting... I don't think they are related to my PR, right?

File: tests/installation/test_chef.py [hunk 2]
```diff
@@ -73,3 +73,34 @@ def test_get_cache_directory_for_link(config):
     )
 
     assert expected == directory
+
+
+def test_no_backslashes_with_file_protocol(config, mocker):
```
Comment (abn) [inline tests/installation/test_chef.py:78]: You should instead probably just updated the existing test case to verify the issue. Like so;

```py
def test_get_cached_archives_for_link(config, mocker):
    chef = Chef(
        config,
        MockEnv(
            marker_env={"interpreter_name": "cpython", "interpreter_version": "3.8.3"}
        ),
    )

    distributions = Path(__file__).parent.parent.joinpath("fixtures/distributions")
    mocker.patch.object(
        chef,
        "get_cache_directory_for_link",
        return_value=distributions,
    )

    archives = chef.get_cached_archives_for_link(
        Link("https://files.python-poetry.org/demo-0.1.0.tar.gz")
    )

    assert archives
    assert set(archives) == {
        Link(path.as_uri())
        for path in distributions.glob("demo-0.1.0*")
    }
```

**Comment by abn** [comment]
> I don't think they are related to my PR, right?
Correct. #2751

File: poetry/installation/chef.py (hunk 1 above)
Comment (fredrikaverpil) [reply poetry/installation/chef.py:84]: Oh, I was googling around for urlencoding this somehow like this, but I failed to find that this was possible with pathlib. I'm starting to like pathlib more and more 😄 👍

File: tests/installation/test_chef.py (hunk 2 above)
Comment (fredrikaverpil) [reply tests/installation/test_chef.py]: Done.

**Pull request merged**